### ISSUE TITLE
aws_iam_policy_attachment not idempotent

### ISSUE THREAD

**Issue by thegranddesign** [hashicorp/terraform#11331]
### Terraform Version
Terraform v0.8.4

### Affected Resource(s)
- aws_iam_policy_attachment

### Description

So what's happening is that I run my terraform config from a clean state (nothing set up) and it sets up everything correctly.  However, if I run it again, when I'd expect it to exit with nothing to do, this happens:

```
~ module.api_server_instance_profile.aws_iam_policy_attachment.policy_to_profile_attachment.2
    roles.#:          "2" => "1"
    roles.2808959860: "api-server" => "api-server"
    roles.397096961:  "bastion" => ""

~ module.bastion_instance_profile.aws_iam_policy_attachment.policy_to_profile_attachment
    roles.#:          "2" => "1"
    roles.2808959860: "api-server" => ""
    roles.397096961:  "bastion" => "bastion"
```

Then I run `apply` again and it does this:

```
~ module.api_server_instance_profile.aws_iam_policy_attachment.policy_to_profile_attachment.2
    roles.#:          "0" => "1"
    roles.2808959860: "" => "api-server"

~ module.bastion_instance_profile.aws_iam_policy_attachment.policy_to_profile_attachment
    roles.#:         "0" => "1"
    roles.397096961: "" => "bastion"
```

Then I run it again and it does the first one.  It just keeps going back and forth like that.

### Debug Output
I've generated the debug output for the `plan` when it should be doing nothing, but instead removes the policy attachment.

I'm not willing to paste this publicly but if you give me an email address to send it to, I have it in a secret gist.

### Expected Behavior
Nothing.

### Actual Behavior
See "Description" above

### Steps to Reproduce
1. `terraform apply`

### Important Factoids
This has been happening for a very long time.  I think since the resource was introduced.

### References
None that I could find

**Comment by grubernaut** [comment]
Hi @thegranddesign, thanks for the issue!

Understandable that for privacy concerns you don't want to paste any debug output. Would it be possible, however, to write a small config file that successfully reproduces the issue? It would be a huge help in debugging the issue further. Thanks!

**Comment by thegranddesign** [comment]
@grubernaut I'm currently under the gun for a deadline.  Can I send the debug log to your email on your Github profile?  Which log level would you prefer?

**Comment by grubernaut** [comment]
@thegranddesign the email in my github profile would work just fine. I'll work on creating a reproduction case tomorrow, but it may also be cause from the configuration. It looks like, from the output pasted, that you're attempting to use `count` with the `aws_iam_policy_attachment` resource, so I'll start there. Thanks!

**Comment by thegranddesign** [comment]
I'm almost certain that that is the issue. I can send you relevant configs as well when I send the trace.

**Comment by grubernaut** [comment]
@thegranddesign just checking in on this again. I believe this issue is directly related to using `count` with the `aws_iam_policy_attachment` resource. Noting that we do have this warning shown on the resource documentation:
> The aws_iam_policy_attachment resource is only meant to be used once for each managed policy. All of the users/roles/groups that a single policy is being attached to should be declared by a single aws_iam_policy_attachment resource.

However, I could see a use case for using `count` if the resource was written as such:
```
variable "policies" { default = ["policy_arn_1", "policy_arn_2"] }

resource "aws_iam_policy_attachment" "test-attach" {
    count = "${length(var.policies)}"
    name = "test-attachment-${count.index}"
    users = ["${aws_iam_user.user.name}"]
    roles = ["${aws_iam_role.role.name}"]
    groups = ["${aws_iam_group.group.name}"]
    policy_arn = "${element(split(",", var.policies), count.index)}"
}
```
I was unable to reproduce the error using the above config though.

**Comment by thegranddesign** [comment]
@grubernaut I'm preparing for a big meeting which is why I haven't gotten you my config yet. Give me just a couple more days. I appreciate you tackling this. 

Yes, that is my use case. The only difference other than my config being much more complicated is that I'm still passing in as a string and splitting on commas. There was a bug in the array passing that kept me from using that for the time being.

**Comment by grubernaut** [comment]
@thegranddesign okay awesome! Thanks

**Comment by thegranddesign** [comment]
@grubernaut sent!  Thank you!

**Comment by grubernaut** [comment]
@thegranddesign awesome, I'll be digging into these shortly. Thanks!

**Comment by grubernaut** [comment]
@thegranddesign I was able to reproduce the error you're experiencing. 
Given the following configuration, and assuming the module at the specified filepath attaches the supplied policies (Preconfigured AWS Policies) to generated iam_roles:
```
module "api_server_instance_profile" {
  source = "./modules/instance_profile"

  name = "api-server"
  policy_ids = "arn:aws:iam::aws:policy/IAMUserChangePassword,arn:aws:iam::aws:policy/service-role/AmazonEC2RoleforSSM,arn:aws:iam::aws:policy/service-role/AWSConfigRole,arn:aws:iam::aws:policy/AmazonAPIGatewa\
yAdministrator,arn:aws:iam::aws:policy/AmazonAPIGatewayInvokeFullAccess,arn:aws:iam::aws:policy/AmazonAppStreamReadOnlyAccess"
  policy_count = 6
}

module "bastion_instance_profile" {
  source = "./modules/instance_profile"
  name = "bastion"
  policy_ids = "arn:aws:iam::aws:policy/IAMUserChangePassword"
  policy_count = 1
}
```

I am able to continuously reproduce the error you are experiencing. However, if I change the list of policy id's passed in the second module declaration, to use policies that were not already attached in the first module's declaration, I can no longer get the issue to reproduce.

Ex:
```
module "api_server_instance_profile" {
  source = "./modules/instance_profile"

  name = "api-server"
  policy_ids = "arn:aws:iam::aws:policy/IAMUserChangePassword,arn:aws:iam::aws:policy/service-role/AmazonEC2RoleforSSM,arn:aws:iam::aws:policy/service-role/AWSConfigRole,arn:aws:iam::aws:policy/AmazonAPIGatewa\
yAdministrator,arn:aws:iam::aws:policy/AmazonAPIGatewayInvokeFullAccess,arn:aws:iam::aws:policy/AmazonAppStreamReadOnlyAccess"
  policy_count = 6
}

module "bastion_instance_profile" {
  source = "./modules/instance_profile"
  name = "bastion"
  policy_ids = "arn:aws:iam::aws:policy/AmazonZocaloFullAccess"
  policy_count = 1
}
```

This is because an `iam_policy` can only have at max 1 policy attachment resource bound to it, as the `iam_policy_attachment` resource will manage every attachment on that policy. This is why we were experiencing the diff between policy attachments for each attachment resource. 

To sum up, in order to fix the issue we need to ensure that each `iam_policy` has only one `iam_policy_attachment` attached to it. This may result in multiple policies that have the same content but different names, or restructuring the policy_attachment resource to maintain the 1-1 relationship between an `iam_policy` and the `iam_policy_attachment` resources.

Closing as not a bug, but happy to discuss further. Thanks!

**Comment by moudixtc** [comment]
@grubernaut It looks like the current way to use `aws_iam_policy_attachment` is to attach every resource that wants to use the policy, but isn't it supposed to be the other way around? For resources like `aws_iam_policy`, I believe what most people would do is to create the policy in a central place, and reference this policy when creating role/group in another terraform configuration. I feel like `aws_iam_policy_attachment` should be declared the same way as `aws_iam_role_policy_attachment` with the only difference that this time the policy attached is a managed policy instead of an inline policy.

**Comment by thegranddesign** [comment]
@grubernaut I've attempted to do things the way you mentioned multiple times but I keep coming back to this.

Basically I think that @moudixtc is right and `aws_iam_policy_attachment` is designed backwards from how most people would use it.

This means that there's simply no way for module authors to allow users of their module to allow their users to pass in policy ARNs and hide the complexity of wiring up the resources the module creates and the policies the users have passed in.

**Comment by catsby** [comment]
Hello! 

Can someone please provide for us a complete reproduction configuration (ideally using modules and no secret things exposed). I'd like to take a better look at what's happening here and hopefully provide additional guidance. 

`aws_iam_policy_attachment` is specifically and intentionally  designed to behave the way it is here. I can see how it may be confusing at first though. It is designed to explicitly control who/what/where a role is attached. It's mission statement could be summed up as "I have a policy, I want it to be attached to these things, and only these things." Having two `aws_iam_policy_attachment` with this mission statement for the same policy naturally conflict. 

From reading above it doesn't sound like that's your intent. If I understand, you want to construct modules to accept a policy as input and attach it to a variable amount of things. For that you need to use `aws_iam_role_policy_attachment`. 

A full example that I can look at and apply ourselves would better help us get on the same page and hopefully provide direction forward. 

Thanks!

**Comment by moudixtc** [comment]
@catsby I can be wrong, but I remember that at the time of my comment, `aws_iam_role_policy_attachment` only supports inline policies, and not managed policies. `aws_iam_role_policy_attachment` now works exactly as I intend to use it. Thanks.

**Comment by thegranddesign** [comment]
I switched as well and it works great for roles.
